### PR TITLE
Implements HTTP Endpoint for Clusters resources

### LINKED ISSUE #966: User management
As user of kubermatic i want to:
- Create Projects
- Create clusters which belong to a project
- Invite other users to my project
- Have different roles for the users in my project (Read, Edit, Owner)

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: Implements HTTP Endpoint for Clusters resources
Fixes#966

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
the old names will be clened up as part of #1251

File: api/pkg/provider/kubernetes/rbac-compliant-cluster.go [hunk 1]
```diff
@@ -109,46 +117,59 @@ func (p *RBACCompliantClusterProvider) New(project *kubermaticapiv1.Project, use
...
-				Name: addon,
-				Cluster: corev1.ObjectReference{
-					Name:       cluster.Name,
-					Namespace:  "",
-					UID:        cluster.UID,
-					APIVersion: cluster.APIVersion,
-					Kind:       "Cluster",
-				},
-			},
+	// TODO: this code deserves refactoring, primarily because we are creating add-ons in a namespace which in fact is created by controller that listens for new clusters.
```
Comment (p0lyn0mial) [inline api/pkg/provider/kubernetes/rbac-compliant-cluster.go:133]: can we move this logic to the controller ?

File: api/pkg/handler/cluster.go [hunk 2]
```diff
@@ -57,11 +64,82 @@ func newClusterEndpoint(sshKeysProvider provider.SSHKeyProvider, cloudProviders
...
+		user := ctx.Value(userCRContextKey).(*kubermaticapiv1.User)
+		clusterProvider := ctx.Value(newClusterProviderContextKey).(provider.NewClusterProvider)
+		project, err := projectProvider.Get(user, req.ProjectName)
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		// validate the request
+		// note that: the ssh keys are optional when provided make sure that they belong to the given project.
+		if len(req.Body.SSHKeys) > 0 {
```
Comment (mrIncompetent) [inline api/pkg/handler/cluster.go:80]: That was already back then a shit approach. 
Can we rather implement a endpoint to assign keys to a cluster?
That way the dashboard does follow-up calls after the create-cluster call & during create-cluster it only sense the cluster object - like a real rest API

I'm fine with keeping it for now but i don't want this to stay in the stable implementation

File: api/pkg/handler/cluster.go [hunk 3]
```diff
@@ -57,11 +64,82 @@ func newClusterEndpoint(sshKeysProvider provider.SSHKeyProvider, cloudProviders
...
+			}
+		}
+		spec := req.Body.Cluster
+		if spec == nil {
+			return nil, errors.NewBadRequest("no cluster spec given")
+		}
+		if err := validation.ValidateCreateClusterSpec(spec, cloudProviders); err != nil {
+			return nil, errors.NewBadRequest("invalid cluster: %v", err)
+		}
+		if spec.Version == "" {
```
Comment (mrIncompetent) [inline api/pkg/handler/cluster.go]: Please fail in this case.
The consume must specify a version - the dashboard already does - so we don't need the defaulting here anymore.
The version is not considered something optional

Comment (mrIncompetent) [reply api/pkg/handler/cluster.go]: Can probably be moved into `validation.ValidateCreateClusterSpec`

File: api/pkg/handler/cluster.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/handler/cluster.go:80]: I will create Issues for that

File: api/pkg/handler/cluster_test.go [hunk 4]
```diff
@@ -8,12 +8,190 @@ import (
 	"testing"
 
 	"github.com/go-test/deep"
+	apiv1 "github.com/kubermatic/kubermatic/api/pkg/api/v1"
 	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
 
+	"strings"
```
Comment (mrIncompetent) [inline api/pkg/handler/cluster_test.go]: should be grouped with other go-sdk imports

File: api/pkg/handler/node_v2.go [hunk 5]
```diff
@@ -37,7 +37,7 @@ const (
 // CreateNodeReqV2 represent a request for specific data to create a node
 // swagger:parameters createClusterNodeV2 createClusterNodeV3
 type CreateNodeReqV2 struct {
-	ClusterReq
+	NewGetClusterReq
```
Comment (mrIncompetent) [inline api/pkg/handler/node_v2.go:40]: Does this break the existing endpoint?

File: api/pkg/handler/node_v2.go [hunk 6]
```diff
@@ -50,7 +50,7 @@ type CreateNodeReqBodyV2 struct {
 // NodeReqV2 represent a request for node specific data
 // swagger:parameters getClusterNodeV2 deleteClusterNodeV2 deleteClusterNodeV3 getClusterNodeV3
 type NodeReqV2 struct {
-	ClusterReq
+	NewGetClusterReq
```
Comment (mrIncompetent) [inline api/pkg/handler/node_v2.go:53]: Does this break the existing endpoint?

File: api/pkg/provider/kubernetes/rbac-compliant-cluster.go [hunk 7]
```diff
@@ -10,12 +10,16 @@ import (
 	kubermaticapiv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
 	machineclientset "github.com/kubermatic/machine-controller/pkg/client/clientset/versioned"
 
+	"time"
+
+	"github.com/golang/glog"
```
Comment (mrIncompetent) [inline api/pkg/provider/kubernetes/rbac-compliant-cluster.go]: grouping is wrong

File: api/pkg/provider/kubernetes/rbac-compliant-cluster.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/provider/kubernetes/rbac-compliant-cluster.go:133]: Yes. We should introduce a `CreateDefaultAddons` functionality into the addon-controller.
But we later need to provide endpoint to dynamically add/remove addons (except the default ones)

File: api/pkg/handler/cluster.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/handler/cluster.go:80]: Issue https://github.com/kubermatic/kubermatic/issues/1372

File: api/pkg/handler/node_v2.go (hunk 5 above)
Comment (p0lyn0mial) [reply api/pkg/handler/node_v2.go:40]: I hope it does not :) I only renamed the struct and added one additional field.

File: api/pkg/provider/kubernetes/rbac-compliant-cluster.go (hunk 7 above)
Comment (p0lyn0mial) [reply api/pkg/provider/kubernetes/rbac-compliant-cluster.go]: thx.

File: api/pkg/handler/cluster_test.go (hunk 4 above)
Comment (p0lyn0mial) [reply api/pkg/handler/cluster_test.go]: okay, thx.

File: api/pkg/handler/cluster.go (hunk 2 above)
Comment (p0lyn0mial) [reply api/pkg/handler/cluster.go:80]: good idea, thanks for creating the issue.

File: api/pkg/provider/kubernetes/rbac-compliant-cluster.go (hunk 1 above)
Comment (p0lyn0mial) [reply api/pkg/provider/kubernetes/rbac-compliant-cluster.go:133]: do we want to create issues for that ?

File: api/pkg/handler/cluster.go (hunk 3 above)
Comment (p0lyn0mial) [reply api/pkg/handler/cluster.go]: okay, moving to `ValidateCreateClusterSpec`

File: api/pkg/provider/kubernetes/rbac-compliant-cluster.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/provider/kubernetes/rbac-compliant-cluster.go:133]: Yes, i'll take care of that

Comment (p0lyn0mial) [reply api/pkg/provider/kubernetes/rbac-compliant-cluster.go:133]: thx.

**Comment by p0lyn0mial** [comment]
@mrIncompetent thanks for the review, comments addressed, PTAL

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**